packages/pyright-internal/src/common/uri/baseUri.ts

```diff
@@ -0,0 +1,288 @@
+/*
+ * baseUri.ts
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ *
+ * Base URI class for storing and manipulating URIs.
+ */
+
+import { some } from '../collectionUtils';
+import { getShortenedFileName, normalizeSlashes } from '../pathUtils';
+import { Uri } from './uri';
+
+export abstract class BaseUri implements Uri {
+    protected constructor(private readonly _key: string) {}
+
+    // Unique key for storing in maps.
+    get key() {
+        return this._key;
+    }
+
+    // Returns the scheme of the URI.
+    abstract get scheme(): string;
+
+    // Returns whether the underlying file system is case sensitive or not.
+    abstract get isCaseSensitive(): boolean;
+
+    // Returns the last segment of the URI, similar to the UNIX basename command.
+    get filename(): string {
+        return this.getBasenameImpl();
+    }
+
+    // Returns the basename without any extensions
+    get withoutExtension(): string {
+        const base = this.filename;
+        const index = base.lastIndexOf('.');
+        if (index > 0) {
+            return base.slice(0, index);
+        } else {
+            return base;
+        }
+    }
+
+    // Returns the extension of the URI, similar to the UNIX extname command.
+    get extname(): string {
+        return this.getExtnameImpl();
+    }
+
+    // Returns a URI where the path just contains the root folder.
+    get root(): Uri {
```

Comment:
is there a reason why some uses `xxxImpl` and others just `abstract`? it looks like there is no common impl on top of `xxxImpl` so making them just `abstract` feels like get same effect?

Reply:
This is a left over side effect of the memoization. The BaseUri would have the public function. It would be cached. The sub classes would implement the real implementation. I only did this for things that were called a whole lot.

I could do like you said, but didn't think it mattered much. Maybe that would be slightly faster.

Reply:
I'll simplify it to match what it would have been without memoization.